core/src/apps/cardano/helpers/cbor_hash_builder.py

```diff
@@ -0,0 +1,62 @@
+from apps.cardano.helpers.lazy_cbor_collection import LazyCborCollection, PauseIteration
+from apps.common import cbor
+
+if False:
+    from typing import Any, Protocol
+
+    class HashFn(Protocol):
+        def update(self, chunk: bytes) -> None:
+            ...
+
+        def digest(self) -> bytes:
+            ...
+
+
+class CborHashBuilder:
+    def __init__(self, hash_fn: HashFn, initial_active_item: LazyCborCollection):
+        self.hash_fn = hash_fn
+        self.active_items_stack: list[LazyCborCollection] = [initial_active_item]
+        self.serializer = cbor.encode_streamed(initial_active_item)
+
+        self.update_hash()
+
+    def update_hash(self) -> None:
+        while (chunk := next(self.serializer)) is not PauseIteration:
+            self.hash_fn.update(chunk)
+
+    def get_hash(self) -> bytes:
+        assert (
+            len(self.active_items_stack) == 1 and self.active_items_stack[0].is_filled()
+        ), "The hash calculation is not finished yet"
+
+        return self.hash_fn.digest()
+
+    def start_lazy_collection(self, collection: LazyCborCollection) -> None:
+        self.add_item_and_update(collection)
+        self.active_items_stack.append(collection)
+
+    def start_lazy_collection_at_key(
```

Comment:
Similarly to `start_lazy_collection`, I would rather use `insert_lazy_collection`, and asserted the current collection is a map.

Reply:
Assertion and comment added and renamed to `add_lazy_collection_at_key` - I hope the name is ok, it seems to me to be more fitting and more descriptive than `insert_lazy_collection`.